### PR TITLE
[Feature]branch-2.2 support hive schema change

### PR DESCRIPTION
**Pull request by adzfolc**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes https://github.com/StarRocks/starrocks/issues/9009

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/external/HiveMetaStoreTableUtils.java [hunk 1]
```diff
@@ -103,4 +110,82 @@ public static long getPartitionStatsRowCount(HiveMetaStoreTableInfo hmsTable, Li
...
+            case "NUMERIC":
+                primitiveType = PrimitiveType.DECIMAL32;
+                break;
+            case "TIMESTAMP":
+                primitiveType = PrimitiveType.DATETIME;
+                break;
+            case "DATE":
+                primitiveType = PrimitiveType.DATE;
+                break;
+            case "STRING":
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/external/HiveMetaStoreTableUtils.java:166]: you may add binary type. user behavior of version 2.2 should be consistent.

Comment (adzfolc) [reply fe/fe-core/src/main/java/com/starrocks/external/HiveMetaStoreTableUtils.java:166]: > you may add binary type. user behavior of version 2.2 should be consistent.

```
            case "STRING":
            case "BINARY":
                return ScalarType.createDefaultString();
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 2]
```diff
@@ -6215,6 +6216,26 @@ public void setHasForbitGlobalDict(String dbName, String tableName, boolean isFo
         }
     }
 
+    public void replayModifyTableColumn(short opCode, ModifyTableColumnOperationLog info) {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:6219]: there is no use of `opCode`, why add this argument?

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 3]
```diff
@@ -6215,6 +6216,26 @@ public void setHasForbitGlobalDict(String dbName, String tableName, boolean isFo
...
+        if (info.getDbName() == null) {
+            return;
+        }
+        String hiveExternalDb = info.getDbName();
+        String hiveExternalTable = info.getTableName();
+        LOG.info("replayModifyTableColumn hiveDb:{},hiveTable:{}", hiveExternalDb, hiveExternalTable);
+        List<Column> columns = info.getColumns();
+        Database db = getDb(hiveExternalDb);
+        HiveTable table;
+        db.readLock();
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:6229]: writeLock?

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 4]
```diff
@@ -7524,6 +7545,26 @@ public void replayUninstallPlugin(PluginInfo pluginInfo) {
...
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:7558]: ditto

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java (hunk 2 above)
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:6219]: why there are two method? `replayModifyHiveTableColumn` and `replayModifyTableColumn`

**Comment by adzfolc** [comment]
PR Regression Link   http://39.101.206.100:8090/view/regression/job/starrocks_pr_regression/614/

File: fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java [hunk 5]
```diff
@@ -177,9 +185,92 @@ public Map<String, HiveColumnStats> getTableLevelColumnStats(List<String> column
...
+        if (db == null) {
+            throw new DdlException("Not found database " + dbName);
+        }
+        db.readLock();
+        try {
+            Table hiveTable = db.getTable(tableName);
+            if (hiveTable == null) {
+                throw new DdlException("Not found table " + dbName + "." + tableName);
+            }
+            ModifyTableColumnOperationLog log = new ModifyTableColumnOperationLog(dbName, tableName, fullSchema);
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:269]: move this out of lock

File: fe/fe-core/src/main/java/com/starrocks/external/hive/Utils.java [hunk 6]
```diff
@@ -143,4 +154,43 @@ public static String getTypeKeyword(String type) {
         }
         return keyword;
     }
+
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/external/hive/Utils.java:157]: add ut for all method added below

File: fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java [hunk 7]
```diff
@@ -177,9 +187,96 @@ public Map<String, HiveColumnStats> getTableLevelColumnStats(List<String> column
...
+
+            if (Catalog.getCurrentCatalog().isMaster()) {
+                ModifyTableColumnOperationLog log = new ModifyTableColumnOperationLog(dbName, tableName, fullSchema);
+                Catalog.getCurrentCatalog().getEditLog().logModifyTableColumn(log);
+            }
+        } finally {
+            db.writeUnlock();
+        }
+
+        ModifyTableColumnOperationLog log = new ModifyTableColumnOperationLog(dbName, tableName, fullSchema);
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:278]: move this into write lock

Comment (adzfolc) [reply fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java]: > move logging into write lock

Moved.

**Comment by adzfolc** [comment]
daily test: http://39.101.206.100:8090/view/Daily/job/master_asan/632/

**Comment by adzfolc** [comment]
run starrocks_fe_unittest

**Review by DorianZheng — APPROVED** [review]
LGTM

**Comment by adzfolc** [comment]
PR Regression Link http://39.101.206.100:8090/view/regression/job/starrocks_pr_regression/662/

**Review by stephen-shelby — APPROVED** [review]

**Comment by adzfolc** [comment]
Branch Release Link: http://39.101.206.100:8090/view/Daily/job/branch_release/890/

**Pull request merged**